# Wikishop  
## Анализ комментариев на токсичность
### Используя модель логистической регрессии и LightGBM поможем выделить "токсичные" комментарии в корпусе текстов


[1. Загрузка библиотек](#data_download)    
[2. Загурзка данных. Предобработка](#data_preprocessing)  
[3. Логистическая регрессия](#logistic)    
[4. LightGBM](#lightGBM)  
[5. Выводы](#Conclusion)    


<a id='data_download'></a>
### 1. Загрузим необходимые библиотеки

In [ ]:
import pandas as pd
import os
import numpy as np


import re
import string

#import torch
#import transformers 

import optuna
import lightgbm as lgb



from sklearn.model_selection import GridSearchCV


import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')
stop_words = stopwords.words('english')

from sklearn.model_selection  import train_test_split
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.feature_extraction.text import TfidfVectorizer

from sklearn.metrics import f1_score, mean_squared_error

import spacy
nlp = spacy.load('en_core_web_sm', disable=['parser','ner'])


[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/juliagrobman/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [ ]:
!python -m spacy download en_core_web_sm

✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')


In [ ]:
!pip install transformers

In [ ]:
pip install nltk


Note: you may need to restart the kernel to use updated packages.


In [ ]:
pip install spacy

Note: you may need to restart the kernel to use updated packages.


<a id='data_preprocessing'></a>
### 2. Загрузка данных. Предобработка

In [ ]:
path_global = '/datasets/toxic_comments.csv'
path_local = '/Users/juliagrobman/Downloads/toxic_comments (1).csv'

if os.path.exists(path=path_global):
    data = pd.read_csv(path_global)
elif os.path.exists(path=path_local):
    data = pd.read_csv(path_local)
else: 
    print('ERROR IN PATH')

In [ ]:
data.head()

text  toxic
0  Explanation\nWhy the edits made under my usern...      0
1  D'aww! He matches this background colour I'm s...      0
2  Hey man, I'm really not trying to edit war. It...      0
3  "\nMore\nI can't make any real suggestions on ...      0
4  You, sir, are my hero. Any chance you remember...      0

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159571 entries, 0 to 159570
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   text    159571 non-null  object
 1   toxic   159571 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 2.4+ MB


In [ ]:
data['text'] = data['text'].astype('U')

### Наш датасет состоят из 160 тыс строк, столбец 'toxic' содержит целевой признак

### Text preprocessing

#### Уберем из текста знаки пунктуации, стоп-слова, пробелы

In [ ]:
def  preprocess_text(text, stopwords):
    # cleaning punctuation
    text = text.translate(str.maketrans('', '', string.punctuation))

    # text to lower
    text = text.lower()

    # removing stopwords
    text = ' '.join([word for word in text.split() if word not in stopwords])

    # removing whitespaces
    text = re.sub(r'\s', ' ', text).strip()

    return text


In [ ]:
def lemmatize_text(text):
    doc = nlp(text)
    return " ".join([token.lemma_ for token in doc])

In [ ]:
data['clean'] = data['text'].apply(lambda x: preprocess_text(x, stopwords=stop_words))

#### Применим лемматизацию

In [187]:
data['lemmatized'] = data['clean'].apply(lambda x: lemmatize_text(x) )

In [188]:
data.head(20)

text  toxic  \
0   Explanation\nWhy the edits made under my usern...      0   
1   D'aww! He matches this background colour I'm s...      0   
2   Hey man, I'm really not trying to edit war. It...      0   
3   "\nMore\nI can't make any real suggestions on ...      0   
4   You, sir, are my hero. Any chance you remember...      0   
5   "\n\nCongratulations from me as well, use the ...      0   
6        COCKSUCKER BEFORE YOU PISS AROUND ON MY WORK      1   
7   Your vandalism to the Matt Shirvington article...      0   
8   Sorry if the word 'nonsense' was offensive to ...      0   
9   alignment on this subject and which are contra...      0   
10  "\nFair use rationale for Image:Wonju.jpg\n\nT...      0   
11  bbq \n\nbe a man and lets discuss it-maybe ove...      0   
12  Hey... what is it..\n@ | talk .\nWhat is it......      1   
13  Before you start throwing accusations and warn...      0   
14  Oh, and the girl above started her arguments w...      0   
15  "\n\nJuelz Santanas Age\n\nIn 2002, Juelz Sant...      0   
16  Bye! \n\nDon't look, come or think of comming ...      1   
17   REDIRECT Talk:Voydan Pop Georgiev- Chernodrinski      0   
18  The Mitsurugi point made no sense - why not ar...      0   
19  Don't mean to bother you \n\nI see that you're...      0   

                                                clean  \
0   explanation edits made username hardcore metal...   
1   daww matches background colour im seemingly st...   
2   hey man im really trying edit war guy constant...   
3   cant make real suggestions improvement wondere...   
4                 sir hero chance remember page thats   
5          congratulations well use tools well · talk   
6                         cocksucker piss around work   
7   vandalism matt shirvington article reverted pl...   
8   sorry word nonsense offensive anyway im intend...   
9                alignment subject contrary dulithgow   
10  fair use rationale imagewonjujpg thanks upload...   
11                 bbq man lets discuss itmaybe phone   
12  hey talk exclusive group wp talibanswho good d...   
13  start throwing accusations warnings lets revie...   
14  oh girl started arguments stuck nose doesnt be...   
15  juelz santanas age 2002 juelz santana 18 years...   
16       bye dont look come think comming back tosser   
17     redirect talkvoydan pop georgiev chernodrinski   
18  mitsurugi point made sense argue include hindi...   
19  dont mean bother see youre writing something r...   

                                           lemmatized  
0   explanation edit make username hardcore metall...  
1   daww match background colour I m seemingly stu...  
2   hey man I m really try edit war guy constantly...  
3   can not make real suggestion improvement wonde...  
4                sir hero chance remember page that s  
5            congratulation well use tool well · talk  
6                         cocksucker piss around work  
7   vandalism matt shirvington article revert plea...  
8   sorry word nonsense offensive anyway I m inten...  
9                alignment subject contrary dulithgow  
10  fair use rationale imagewonjujpg thank upload ...  
11                  bbq man let discuss itmaybe phone  
12  hey talk exclusive group wp talibanswho good d...  
13  start throw accusation warning let review edit...  
14  oh girl start argument stick nose do not belon...  
15  juelz santanas age 2002 juelz santana 18 year ...  
16       bye do not look come think comme back tosser  
17     redirect talkvoydan pop georgiev chernodrinski  
18  mitsurugi point make sense argue include hindi...  
19  do not mean bother see you re write something ...

### Разделим данные на обучающую, валидационную и тестовую выборки

In [189]:
features = data['lemmatized']
target = data['toxic']

In [190]:
features_train_full, features_valid, target_train_full, target_valid = train_test_split(features, target, test_size=0.2, shuffle=True)
features_train, features_test, target_train, target_test = train_test_split(features_train_full, target_train_full, test_size=0.2, shuffle=True)

#### Применим if-idf для векторизации текстов

In [191]:
count_tf_idf = TfidfVectorizer()
features_train = count_tf_idf.fit_transform(features_train)

In [192]:
features_valid = count_tf_idf.transform(features_valid)
features_test = count_tf_idf.transform(features_test)

<a id='logistic'></a>
### 3. Логистическая регрессия

#### Обучим логистическую регрессию с гиперпараметрами

In [193]:
params = {
    'penalty' : ['l1','l2'],
    'C': np.logspace(-3,3,7),
    'solver':['lbfg', 'liblinear']
}

model = LogisticRegression(random_state=12345)
grid = GridSearchCV(model, param_grid=params, scoring='f1')
grid.fit(features_train, target_train)
predictions = grid.predict(features_valid)
score = f1_score(predictions, target_valid)
score

/opt/anaconda3/lib/python3.8/site-packages/sklearn/model_selection/_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "/opt/anaconda3/lib/python3.8/site-packages/sklearn/model_selection/_validation.py", line 531, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py", line 1304, in fit
    solver = _check_solver(self.solver, self.penalty, self.dual)
  File "/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py", line 433, in _check_solver
    raise ValueError("Logistic Regression supports only solvers in %s, got"
ValueError: Logistic Regression supports only solvers in ['liblinear', 'newton-cg', 'lbfgs', 'sag', 'saga'], got lbfg.

  warnings.warn("Estimator fit failed. The score on this train-test"


In [ ]:
grid.best_params_

{'C': 10.0, 'penalty': 'l1', 'solver': 'liblinear'}

In [ ]:
# Проверим на тестовой выборке:
predictions = grid.predict(features_test)


In [ ]:
score_test = f1_score(target_test, predictions)
print('F1 score {:.2f}'.format(score_test))

F1 score 0.79


#### Попробуем проверить нашу модель на "токсичном" и "нетоксичном" предложении

In [ ]:
test_sent = 'look at yourself, you are a fat, black, sodding idiot, thinking just about money'

In [ ]:
test_sent_0 = 'I will surely recommend this product to my family and friends, thanks!'

In [ ]:
test_sent = preprocess_text(test_sent, stop_words)
test_sent = lemmatize_text(test_sent)

In [ ]:
test_sent_0 = preprocess_text(test_sent_0, stop_words)
test_sent_0 = lemmatize_text(test_sent_0)

In [ ]:
test_sent_0

'surely recommend product family friend thank'

In [ ]:
test = count_tf_idf.transform([test_sent])

In [ ]:
test_0 = count_tf_idf.transform([test_sent_0])

In [ ]:
grid.predict(test)

array([1])

In [ ]:
grid.predict(test_0)

array([0])

### Наша модель хорошо предсказывает токсичные комментарии, f1 выше трешхолда в 0.75

<a id='lightGBM'></a>
### 4.Попробуем другую модель: LightGBM

In [ ]:
## Попробуем LightGBM

train_data  = lgb.Dataset(features_train, label=target_train)
test_data = lgb.Dataset(features_test, label=target_test)

parameters = {
    'application': 'binary',
    'objective': 'binary',
    'metric' : 'auc',
    'boosting': 'dart',
    'num_leaves': 50,
    'feature_fraction': 0.5,
    'bagging_fraction': 0.5,
    'bagging_freq': 20,
    'learning_rate': 0.05,
    'verbose': 0
}

classifier = lgb.train(
    parameters,
    train_data,
    valid_sets=test_data,
    num_boost_round=1000,
    early_stopping_rounds=100
    
)
predict = classifier.predict(features_valid)


/opt/anaconda3/lib/python3.8/site-packages/lightgbm/engine.py:181: UserWarning: 'early_stopping_rounds' argument is deprecated and will be removed in a future release of LightGBM. Pass 'early_stopping()' callback via 'callbacks' argument instead.
  _log_warning("'early_stopping_rounds' argument is deprecated and will be removed in a future release of LightGBM. "


[LightGBM] [Warning] objective is set=binary, application=binary will be ignored. Current value: objective=binary
[LightGBM] [Warning] objective is set=binary, application=binary will be ignored. Current value: objective=binary
[LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.610585 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Warning] objective is set=binary, application=binary will be ignored. Current value: objective=binary
[1]	valid_0's auc: 0.795793
[2]	valid_0's auc: 0.860647
[3]	valid_0's auc: 0.88404


/opt/anaconda3/lib/python3.8/site-packages/lightgbm/callback.py:223: UserWarning: Early stopping is not available in dart mode
  _log_warning('Early stopping is not available in dart mode')


[4]	valid_0's auc: 0.916159
[5]	valid_0's auc: 0.924503
[6]	valid_0's auc: 0.928224
[7]	valid_0's auc: 0.929954
[8]	valid_0's auc: 0.931208
[9]	valid_0's auc: 0.932753
[10]	valid_0's auc: 0.93401
[11]	valid_0's auc: 0.935002
[12]	valid_0's auc: 0.935576
[13]	valid_0's auc: 0.936272
[14]	valid_0's auc: 0.937503
[15]	valid_0's auc: 0.938857
[16]	valid_0's auc: 0.939691
[17]	valid_0's auc: 0.939805
[18]	valid_0's auc: 0.940655
[19]	valid_0's auc: 0.942552
[20]	valid_0's auc: 0.942944
[21]	valid_0's auc: 0.944051
[22]	valid_0's auc: 0.945197
[23]	valid_0's auc: 0.94531
[24]	valid_0's auc: 0.946483
[25]	valid_0's auc: 0.946751
[26]	valid_0's auc: 0.946826
[27]	valid_0's auc: 0.947053
[28]	valid_0's auc: 0.947024
[29]	valid_0's auc: 0.947747
[30]	valid_0's auc: 0.948007
[31]	valid_0's auc: 0.948325
[32]	valid_0's auc: 0.94851
[33]	valid_0's auc: 0.948687
[34]	valid_0's auc: 0.948937
[35]	valid_0's auc: 0.949004
[36]	valid_0's auc: 0.949113
[37]	valid_0's auc: 0.949783
[38]	valid_0's auc: 0.9

In [ ]:
pred = [1  if el > 0.4 else 0 for el in list(predict)]

In [ ]:
score_light = f1_score(target_valid, pred)

In [ ]:
print('F1 score по LightGBM {:.2f}'.format(score_light))

F1 score по LightGBM 0.78


In [ ]:
pred_test = classifier.predict(features_test)

In [ ]:
def best_f1_score
pred_test_light =  [1  if el > 0.4 else 0 for el in list(pred_test)]

In [ ]:
# попробуем LigthGBM на тестовой выборке
score_light_test = f1_score(target_test, pred_test_light)
score_light_test

0.7985777034093287

In [ ]:
print('F1 score на тестовой выборке по LightGBM {:.2f}'.format(score_light_test))

F1 score на тестовой выборке по LightGBM 0.80


### Получилось на тестовой выборке немного улучшить показатели логистической регрессии, F1 мера составляет чуть менее 0.8

<a id='Conclusion'></a>
### 5. Выводы
* Зазрузили данные с необработанными комментариями
* Провели предобработку текстов: 
    - убрали знаки пунктуации, стоп-слова, привели к нижнему регистру
    - провели лемматизацию
    - с помощью if-idf провели векторизацию
* Обучили модели Логистической регрессии и LightGBM с различными гиперпараметрами
* Добились f1-score на тестовой выборке 0.80  
* Проверили модели на токсчичном и нетоксичном предложениях, убедились, что модель адекватно предсказывает "токсчичность"